REF: pandas/io/formats/format.py

ivanovmg opened the pull request:
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Partially addresses https://github.com/pandas-dev/pandas/issues/36407

Before splitting ``DataFrameFormatter`` into more dedicated classes, I decided to refactor the class itself, to make the outstanding refactoring more manageable.

As suggested by @jreback, I made this refactor small, trying to split big functions into smaller ones and find some better naming.
Yet, rather small changes are done so far, just to keep diffs readable.
Once approved, I will move on to further refactor.

ivanovmg committed:
REF: extract formatters property

ivanovmg committed:
REF: extract justify property

ivanovmg committed:
REF: extract columns property

ivanovmg committed:
REF: extract col_space property

ivanovmg committed:
REF: extract sparsify property

ivanovmg committed:
REF: extract max_rows_displayed property

ivanovmg committed:
CLN: eliminate 'w' and 'h' attributes

ivanovmg committed:
TYP: annotate col_space

ivanovmg committed:
TYP: annotate formatters

ivanovmg committed:
TYP: annotate columns

ivanovmg committed:
TYP: annotate max_rows_displayed

ivanovmg committed:
TYP: add formatters property in superclass

To ensure that mypy does not raise override error

ivanovmg committed:
TYP: add columns property in superclass

To ensure that mypy does not raise override error

ivanovmg committed:
REF: extract method _is_in_terminal

ivanovmg committed:
REF: extract max_cols_adj property

ivanovmg committed:
REF: extract max_rows_adj property

ivanovmg committed:
REF: extract _is_screen_narrow/short methods

ivanovmg committed:
REF: simplify logic in max_rows_adj

ivanovmg committed:
REF: check for terminal in max_cols_adj

This is done for consistency with max_rows_adj

ivanovmg committed:
REF: extract truncate_h/v and is_truncated props

ivanovmg committed:
REF: truncate_h -> is_truncated_horizontally

ivanovmg committed:
REF: truncate_v -> is_truncated_vertically

ivanovmg committed:
TYP: make is_truncated compatible with superclass

ivanovmg committed:
CLN: remove unnecessary variables

ivanovmg committed:
REF: extract _get_number_of_auxillary_rows

ivanovmg committed:
REF: use var header_row for clarity

ivanovmg committed:
TYP: ignore assignment error (setters)

ivanovmg committed:
TYP: specify max_rows type

pep8speaks commented:
Hello @ivanovmg! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/io/formats/html.py`](https://github.com/pandas-dev/pandas/blob/6f1a81f45cad8c0d9f20e12f7059d30b0d48157f/pandas/io/formats/html.py):

> [Line 459:89](https://github.com/pandas-dev/pandas/blob/6f1a81f45cad8c0d9f20e12f7059d30b0d48157f/pandas/io/formats/html.py#L459): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 88 characters)

ivanovmg committed:
LINT: long comment line

simonjayhawkins commented:
Thanks @ivanovmg for the PR. General Question.

DataFrameFormatter is not public. Is the validation always necessary or only during \_\_init__? Would _validator_foo() validators (perhaps even in attrs style https://www.attrs.org/en/stable/examples.html#validators) be more appropriate than property setters.

ivanovmg commented:
> 
> 
> Thanks @ivanovmg for the PR. General Question.
> 
> DataFrameFormatter is not public. Is the validation always necessary or only during __init__? Would _validator_foo() validators (perhaps even in attrs style https://www.attrs.org/en/stable/examples.html#validators) be more appropriate than property setters.

My intention was indeed to extract parameters validation/conversion. It is not only validation. I will need to carefully look through the code to see if there is any assignment to the attributes via the setters defined. Probably these setters are used only once here.

I think that if we potentially derive ``ConsoleFormatter``, ``HTMLFormatter`` and ``LatexFormatter`` from ``DataFrameFormatter``, these properties with setters will be useful, since they will make initialization neat (just self.some_attr = some_attr). One the other hand it may be not very clear that there is a setter hiding behind.

jbrockmendel commented on pandas/io/formats/format.py:
```diff
@@ -451,9 +453,27 @@ def get_adjustment() -> TextAdjustment:
+
+    @property
+    def is_truncated(self) -> bool:
+        self._is_truncated: bool
```
what is this line for?

ivanovmg replied:
This is necessary to make mypy happy. For some reason it thinks that there is a violation of Liskov substitution principle even if is_truncated property (the same name as the attribute in superclass) in the subclass has the same return type. Therefore, I had to create the same property signature in the superclass.

jbrockmendel replied:
could you put `_is_truncated: bool` at the class level just before `__init__`?

ivanovmg replied:
This is certainly possible. But considering questionable use of property setters (like discussed with @simonjayhawkins and which you imply in your other comment), probably I will just create initialization methods for each attribute where required. Therefore, there will be no need in the constructs you refer to.

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,71 +583,208 @@ def __init__(
+        """Number of columns fitting the screen."""
+        self._max_cols_adj: Optional[int]
+
+        try:
```
is this new?

jreback replied:
i c was a hasattr check before (can you restore). I suspect once you change the class structure this won't be needed anyhow (or rather it will be used / needed in some of the classes but not all)

ivanovmg replied:
@jreback, if you inquire about try/except construction, there there is a reason for it.
There is an assignment to max_cols_adj at line 974. After that assignment I want only that value to be used. Therefore, in the property I want to return ``self._max_cols_adj`` right away if it exists.
Technically, this can be refactored so that the setter takes care of ``self._max_cols_adj`` value.

Yes, this property will be needed only for console and html outputs.
I intended to make classes separation afterwards.
Do you still suggest that I restore the previous implementation? Because there was too much going on in the ``_chk_truncate`` method.

ivanovmg replied:
Or do you suggest that I use hasattr here instead of try/except? I would probably better go for a setter for more clarity.

ivanovmg committed:
REF: use hasattr instead of try on AttributeError

ivanovmg committed:
REF: extract helper truncate methods

This simplifies method _chk_truncate.
New methods extracted:
 - _truncate_vertically
 - _truncate_horizontally

ivanovmg committed:
CLN: remove unnecessary if statement

ivanovmg committed:
CLN: remove unnecessary else clause

ivanovmg committed:
CLN: move str_index to where it is used

ivanovmg committed:
CLN: several assignments

ivanovmg committed:
REF: extract _get_strcols method

This returns strcols before any truncation.
It will be useful for LaTeX formatting, for instance,
where no truncation is present.
Should consider polymorphism later (with and without truncation).

ivanovmg committed:
CLN: remove stringified variable

ivanovmg committed:
REF: extract _get_strcols_without_index method

ivanovmg committed:
REF: extract _truncate_strcols method

ivanovmg committed:
REF: extract _get_string_representation

This method gets string representation of the table
to be printed or saved.

ivanovmg committed:
CLN: simplify logic in _get_string_representation

ivanovmg committed:
CLN: simplify condition to wrap around

ivanovmg committed:
REF: extract _fit_strcols_to_terminal_width

ivanovmg committed:
CLN: better names for start, end

ivanovmg committed:
REF: get rid of magic number 3

Here 3 means the number of dimension info lines
(two new lines and one actual info line).
This commit makes it explicitly.

ivanovmg commented:
I couldn't help but extracting some more methods. I hope the diffs are is still readable. These changes will be helpful for the upcoming polymorphism updates (formatters with and without truncation).

ivanovmg committed:
REF: incrementally add numbers of auxillary rows

ivanovmg committed:
CLN: rename _chk_truncate -> _truncate

This is more reasonable name for what it does.

ivanovmg committed:
REF: _truncate_horizontally without concatenation

ivanovmg committed:
CLN: simplify logic in _truncate_horizontally

ivanovmg committed:
CLN: simplify logic in _truncate_vertically

ivanovmg committed:
REF: _truncate_vertically without concatenation

ivanovmg committed:
LINT: list comprehension

ivanovmg committed:
LINT: remove extra concat import

ivanovmg committed:
TYP: remove type casting

ivanovmg committed:
REF: extract _insert_dot_separator_vertical method

ivanovmg committed:
REF: extract _insert_dot_separator_horizontal

ivanovmg committed:
CLN: _truncate_strcols -> _insert_dot_separators

After previous methods extractions it became more clear
that this method actually only inserts dots
to the already truncated strcols.

ivanovmg committed:
TYP: add annotations on new methods

ivanovmg committed:
TYP: replace casting with asserts

ivanovmg commented:
Something is wrong with the CI. The environment is not getting built time after time.

jbrockmendel commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
+        return self._justify
+
+    @justify.setter
+    def justify(self, justify):
```
are these properties set dynamically (i.e. after `__init__`?)

ivanovmg replied:
No, turns out that they are used only in ``__init__``.

jreback replied:
yeah i would be in favor of simply typing these on the class. i am not sure why that is causing issues; we do this in lots of places and its the idiomatic soln. (for something you are setting that is different), but if you are *only* setting it in ``__init__`` then prefer to not have these accessors

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
+        self.adj = get_adjustment()
+
+    @property
+    def sparsify(self):
```
can you type

ivanovmg replied:
Done.

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
-            self.formatters = {}
-        elif len(frame.columns) == len(formatters) or isinstance(formatters, dict):
-            self.formatters = formatters
+            self._formatters = {}
```
i think if you type this it might be able to remove some of the mypy errors above (alt is to put in the class as typed)

ivanovmg replied:
I replaced setters with initialization methods for each parameter. Now there is no problem with mypy.

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
+        self._col_space: ColspaceType
         if col_space is None:
-            self.col_space = {}
+            self._col_space = {}
```
this likey needs typing

ivanovmg replied:
Added

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
+        return bool(self.max_cols == 0 or self.max_rows == 0)
+
+    @property
+    def max_cols_adj(self) -> Optional[int]:
```
max_cols_fittted maybe?

ivanovmg replied:
Replaced as suggested

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
 
-        self.bold_rows = bold_rows
-        self.escape = escape
+    def _is_screen_narrow(self, max_width) -> bool:
```
is this private for a reason?

ivanovmg replied:
I thought that this method is not going to be used anywhere outside of the class. So, I made this private. I think some other methods also should be private, but I left them as is for now. Unless you have other thoughts on the topic.

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
-        the frame up.
-        """
-        from pandas.core.reshape.concat import concat
+    def _get_number_of_auxillary_rows(self) -> int:
```
doc-string

jreback replied:
as its not obvious what auxillary is (reading down it becomes clear but a doc-string is good)

ivanovmg replied:
Did that.

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
-    def _to_str_columns(self) -> List[List[str]]:
+        return num_rows
+
+    def _is_screen_short(self, max_height) -> bool:
```
can you group with is_screen_narrow

ivanovmg replied:
Placed next to each other

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
+            self._truncate_vertically()
+
+    def _truncate_horizontally(self) -> None:
+        assert self.max_cols_adj is not None
```
doc-string if y ou can

ivanovmg replied:
Done

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
+        self.tr_col_num = col_num
+
+    def _truncate_vertically(self) -> None:
+        assert self.max_rows_adj is not None
```
same

ivanovmg replied:
Documented

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
-
         if not is_list_like(self.header) and not self.header:
-            stringified = []
+            strcols = []
```
type

ivanovmg replied:
Done

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
-                fmt_values = _make_fixed_width(
-                    fmt_values, self.justify, minimum=header_colwidth, adj=self.adj
-                )
+        strcols = []
```
type

ivanovmg replied:
Done

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +583,446 @@ def __init__(
+
+        return strcols
+
+    def _insert_dot_separators(self, strcols: List[List[str]]) -> List[List[str]]:
```
you could define List[List[str]] at the top as an alias, maybe Screen, see if that is more clear

jreback replied:
woudln't rename strcols -> screen if you can do that

ivanovmg replied:
I think that it is more clear that strcols is a a list of columns. If we rename it to screen, then that would not be clear. I think such rename may be more useful for the list of rows instead.

jreback reviewed (changes requested):
wow this is a lot cleaned up. some comments.

ivanovmg committed:
DOC: docstring to _get_number_of_auxillary_rows

ivanovmg committed:
CLN: place _is_screen_narrow/short funcs together

ivanovmg committed:
DOC: docstrings for _truncate_horizontally/vert

ivanovmg committed:
TYP: add for strcols, justify and sparsify

ivanovmg committed:
REF: replace sparsify setter with initialization

ivanovmg committed:
REF: replace formatters setter with initialization

ivanovmg committed:
REF: replace col_space setter with initialization

ivanovmg committed:
REF: replace columns setter with initialization

ivanovmg committed:
REF: replace justify setter with initialization

ivanovmg committed:
CLN: reorder terminal/screen related methods

ivanovmg committed:
TYP: replace casting with actual int conversion

ivanovmg committed:
TYP: assert instead of casting

ivanovmg committed:
CLN: max_rows/cols_adj -> max_rows/cols_fitted

ivanovmg committed:
TYP: move _is_truncated type declaration on top

simonjayhawkins commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +569,414 @@ def __init__(
+            return strcols
+
+        if is_list_like(self.header):
+            assert isinstance(self.header, list)
```
is_list_like is not the same as isinstance(..., list). can you revert

ivanovmg replied:
Good catch. Thank you

simonjayhawkins commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +569,414 @@ def __init__(
 
-        self.bold_rows = bold_rows
-        self.escape = escape
+        if hasattr(self, "_max_cols_fitted"):
```
hasattr checks don't play play well with mypy

max_rows_displayed is called from is_truncated_horizontally which is quite prevalent.

optimisation here may be premature here, but in I think generally avoiding hasattr checks is preferable is called often.

```
class MyClass:
    pass

cls = MyClass()

%timeit hasattr(cls, "_max_cols_fitted")
# 120 ns ± 8.08 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

class MyClass:
    _max_cols_fitted = None

cls = MyClass()

%timeit cls._max_cols_fitted is not None
# 85.5 ns ± 1.94 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

what is the reason for changing `max_rows_adj` to a property?

ivanovmg replied:
The reason for making the property out of ``max_rows_adj/max_rows_fitted`` was to place the computation inside the function. Before that it was scattered in the other methods, while it was difficult to understand how the calculation was actually carried out. The present implementation makes it more clear.

Before hasattr check I used try/except AttributeError check. But if I understood correctly, @jreback suggested to revert to hasattr check, which was there originally.

simonjayhawkins replied:
having the separate function is good. maybe _calc_max_cols_fitted and just call it once and assign to attribute like before

> Before hasattr check I used try/except AttributeError check. But if I understood correctly, @jreback suggested to revert to hasattr check, which was there originally.

it was only called once before.

ivanovmg replied:
If performance is a concern, we can consider ``cached_property`` decorator or it's equivalent created for Python 3.7. I can do this if required.

ivanovmg replied:
> 
> 
> having the separate function is good. maybe _calc_max_cols_fitted and just call it once and assign to attribute like before
> 
> > Before hasattr check I used try/except AttributeError check. But if I understood correctly, @jreback suggested to revert to hasattr check, which was there originally.
> 
> it was only called once before.

This is fine with me. I will change as you suggest.

simonjayhawkins replied:
> If performance is a concern

it's not. just suggesting avoiding certain patterns

from https://github.com/python/mypy/issues/1424#issuecomment-223795437

> In general i despise hasattr() checks; it's usually much better to add a
class level initialization to None and check for that.

simonjayhawkins replied:
> we can consider `cached_property` decorator or it's equivalent created for Python 3.7. I can do this if required.

This is not in the standard library til 3.8?

ivanovmg replied:
It is in the standard library starting from 3.8. Before that there wasn't one. So, I was thinking could b reasonable to implement the cached property for the needs, if the version < 3.8.

ivanovmg committed:
Revert "TYP: assert instead of casting"

This reverts commit dcb2b7003fb7bb8ab29be7f9fa8307930e4afb0a.

ivanovmg committed:
REF: assign to max_cols/rows_fitted at init

This allows one avoid multiple calls of the properties,
which were there originally.

jreback commented:
we already have a cached property decorator

cache_readonly

though only use it if we are actually doing something non trivial

simonjayhawkins commented:
yeah cache_readonly is in cython without a stub file, so using here would undo all the typing benefits since properties would resolve to 'Any'

jreback commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +569,409 @@ def __init__(
-        """
-        Render a DataFrame to a list of columns (as lists of strings).
+    def _truncate_horizontally(self) -> None:
+        """Remove columns, which are not to be displayed and adjust formatters.
```
wouldn't be against making this a non-mutating function and instead just return the new values (and set where this is called), but you did explain so ok.

jreback reviewed (approved):
lgtm. ping on green. nice improvement. happy to have additional to separate classes as you mentioned.

jreback commented:
thanks @ivanovmg keep em coming!

jorisvandenbossche commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +569,409 @@ def __init__(
+                x
+                for x in range(self.frame.shape[0])
+                if x < row_num or x >= len(self.frame) - row_num
+            ]
```
I suppose this is the reason for the slowdown (https://github.com/pandas-dev/pandas/issues/36636). In the original code, we do 2 slices + a concat. What was the motivation for this (slow) for loop? To select the subset in a single go?

ivanovmg replied:
Right, it was supposed to be more readable to extract necessary rows. But, apparently, I never considered cases with large number or rows. Will fix it.

jorisvandenbossche commented on pandas/io/formats/format.py:
```diff
@@ -565,315 +569,409 @@ def __init__(
+            - tr_row_num
+        """
+        assert self.max_rows_fitted is not None
+        row_num = self.max_rows_fitted // 2
```
BTW, I find the "max_rows_fitted" naming instead of "max_rows_adjusted" as it was before a bit confusing, as the "fitting" seems to come from "fit on the screen", but the eventual max_rows are not necessarily what fit on the screen, that depends on your settings.

ivanovmg replied:
OK. Will change in the upcoming PRs.
